### DISCUSSION TITLE
Option to show average, not sum

### DISCUSSION THREAD

**Discussion by seekM** [iSoron/uhabits#1299] [Feature Requests]
Hi,

when you create a new habit you have the option to go for "yes / no" or "measurable". If you choose "measurable" and start tracking, at the section "history" (or "progress"? hope it's clear what I mean, I'm using the german language variant) you can select an interval (e.g. "week" or "month") and can see for each e.g. week or month the sum of the values entered in this interval.

However, I'm not interested in the sum for the choosen interval, but for the average. Example use case: I'm using this app to track my weight.

Is this already possible? I couldn't figure out how to do this. If not, it would be great if there was on option to choose what you are interseted in (sum or average).

Thank you for offering this app!

**Comment by neekubee** [comment]
Use case: Weight Tracking

The goal is to gain/lose x units in y number of days. The app as I know it doesn't understand any similar use cases. If one is to add x units, it would simply add up, which the user doesn't want for weight tracking. What the user expects is to track how much increase or decrease, regarding their weight, over a period of time.

Maybe you could try to manually calculate the difference and that's what you input. Say your goal is to lose x units in 30 days and today you input 1 unit, that would mean you're nearer your goal. But if it decreases 1 unit, that would mean farther.

How would you know how accurate the data is? Maybe add another habit for actual weight but using the calculated values for the progress.

I dont understand why you would want to know what your average weight is for a given period. Why?

**Reply by neekubee** [reply]
Anyway, I also want to know the average of some of the things I track but I think the score is a better indicator.

**Reply by seekM** [reply]
My goal is not to gain/lose x units in y number of days. My goal is to track my weight over some period of time for health reasons. If you notice that your weight changes even though you don't eat differently or don't do different activties, it can be a sign that there might be a health problem. Therefore it would be nice to see the average (and not the sum) of the weight entered if I select e.g. "month" at the "progress/history" diagramm.

**Reply by vyda** [reply]
There are also use cases where you have no fixed goal. Then it does not make sense to track a difference.

Use case: track development of a measurable ability

E.g. when I measure how many pushups I can do in one go (or burpees or squats ...) and track that number.

I would like to see how this ability changes over time, so an average is a good indicator. The total number of pushups is not interesting to me because that is a function of how often I test myself.

**Reply by herrbrixter** [reply]
This idea is different from what this post proposes. Please open a different thread.

**Comment by MarkJeronimus** [comment]
I too would like to have an 'average' for habits, but I imagine it as a different habit type.

To elaborate, current Measurable habits add up numbers for week, month, year. So if you take 2 pills per day it's 14 pills per week, etc.

The new "Average Measurable" habit will not add up numbers but average them instead. So if I weigh 152 lbs and 154 lbs on two days of a week (and didn't measure myself on other days) it should how the week "total" as 153 lbs instead of the current 306 lbs.

**Comment by djvanderlaan** [comment]
In my case I am tracking sleep quality; I score this from 1 to 10. For me a total score over a week or month is not that interesting. I am more interested in the average sleep quality. 

But also for other stuff like the amount of water drank during the day, the average might be more useful than the total. 

One thing that needs to be thought about is how to handle the situation where no value is given for a given day. Personally I would prefer if these days are ignored when calculating the average. So if I had a score of 4 on monday, 6 on tuesday, forgot to fill in the score from wednesday to saturday and had a score of 8 on sunday, the average would be 6 for that week.

For other scores like the number of push ups done during the day, it might make more sense to assume that a missing value means that no pushups were done. So for calculating the average these would count as 0. 

So it would be nice to have the option to choose how missing values are handled. Although the default should be to ignore those: a user can always fill in a 0.

**Reply by Janssenhidal** [reply]
Hi @djvanderlaan,

I saw your suggestion about adding averages in Habit App to better track things like sleep and weight. I have been waiting for this for a while, so I decided to take a stab at it. My coding skills are basic, but I think I can handle this part.

Before I proceed, could you confirm how you would like the averages to work?

Weekly average: last 7 days rolling, or the current calendar week based on the app’s “week starts on” setting?

Monthly average: last 30/31 days rolling, or the current calendar month?

Yearly average: last 365 days rolling, or the current calendar year?

Empty days: exclude empty entries from the divisor, correct?

Once I have your preference, I will **try** to implement it accordingly. Although personally, I would rather have rolling windows to avoid arbitrary boundaries.

**Reply by xdevs23** [reply]
Chiming in to give my two cents nobody asked for:

IMO in the History view that currently shows statistics summed up, averages should be for the calendar month and not rolling.

Empty days should be configurable. Either exclude or zero.

The rolling windows don't really make a lot of sense in the History view as they are grouped per month. But this could be modified to accomodate an option that allows switching to a relative view and use rolling windows which would show -1m, -2m, -3m etc.

**Reply by Janssenhidal** [reply]
Don't worry, appreciate you chiming in.

I totally see your point about the History view being grouped per month, that definitely makes calendar-based averages more natural there. At the same time, if someone is using the app to track things like weight, a rolling view might make more sense. For example, if your week starts on Monday and you want to check your “last week” average on a Tuesday, you’d only get a 2-day average instead of a proper 7-day snapshot.

That’s why I was thinking something like last 7 / 30 / 90 / 365 days might be a more useful option.

Do you by any chance have a platform where it’s easier to go back and forth on this? I’d love to get more of your opinion as I try to shape how this should work best.

**Reply by djvanderlaan** [reply]
Thanks for picking this up. 

> Weekly average: last 7 days rolling, or the current calendar week based on the app’s “week starts on” setting?
> Monthly average: last 30/31 days rolling, or the current calendar month?
> Yearly average: last 365 days rolling, or the current calendar year?

No strong preference, but probably calender weeks/months/years. As @xdevs23 mentioned this also fits better with the current user interface. The 'advantage' of averages instead of sums, is that, in your example, you can calculate an average weight of current month even in the first week of that month (as long as you have only 1 measurement that month) and compare that to the average of the previous month.  At least the figures are comparable (although the partial month score of course will be less precise).

> Empty days: exclude empty entries from the divisor, correct?

Yeah, that would be my preference. Otherwise the current sum would also work (although it would be nice if that sum could be translated to, for example, per week or per day).

**Reply by Janssenhidal** [reply]
Noted... those are valid points. I’d love some help/feedback on the UI design side:

Should averages be integrated into the existing target card (e.g., a dropdown to switch between Total/Average)?

Or would it be better to create a separate card with its own chart for averages?

I’d really appreciate any ideas or suggestions from you guys on what would fit best with the current design.

Thanks!

**Reply by herrbrixter** [reply]
> Weekly average: last 7 days rolling, or the current calendar week based on the app’s “week starts on” setting?

Time span should be fixed otherwise it would be hell to compare without a stable base.
Week - based on "first day of the week" setting.
Month - calendar month

> Empty days: exclude empty entries from the divisor, correct?

Please include the option to treat empty days as 0 and include in the divisor.

Use case: I want to run average of 3km daily in a week

Sunday - Missing
Monday - 2km
Tuesday - Missing
Wednesday - Missing
Thursday - 4km
Friday - Missing
Saturday - Missing

Average daily run for the week is 6/7 = 0.86km and not 6/2 = 3km

**Comment by xdevs23** [comment]
I use the habit tracker to track more general things, not just habits. I find the app extremely easy to use and well suited for long time tracking, and I genuinely enjoy using it. The only thing missing for me is average tracking.

I use the app to track:
* Hours of sleep
* Mood
* Aggregated bank account balances
* Aggregated savings
* Weight
among other things

I enter the data every day. Summing the bank account balances of every day for a month results in hard to read numbers and graphs. I would like to see: how much money did I have in average? Sum of all entries divided by number of days data has been entered. Allow for treating no data as zero.

Same for hours of sleep. I don't really care how many hours I slept in total, rather I want to know the average across a month to determine if I need to change something. Sure, I could run the calculations myself but it's just a matter of convenience.

Weight is the one that makes the most sense for this use case. You don't want to know the sum of weights, you want to know how much you weigh in average so you can compare weights across days, weeks, months and more. Sure, I could again, sum it up and divide manually. Or I could track deltas instead. However, the app does not support negative numbers so I can't really use a baseline.

**Comment by MarkJeronimus** [comment]
I just updated uHabits from the F-Droid store (for some reason it was the only app not auto-updating) and found that instead of this 'average' feature, it added a fancy confetti effect upon entering data. Why is a bling feature prioritized over a QOL feature? This makes me question the sanity of the authors somewhat. It can't even be disabled and it's ridiculous for "bad habits" like how much alcohol I drank.

**Comment by Janssenhidal** [comment]
PR has been submitted regarding this topic. Awaiting review.

https://github.com/iSoron/uhabits/pull/2215

**Reply by djvanderlaan** [reply]
Cool! Thanks. Unfortunately I can't really judge the code changes. The screenshots look good.

**Comment by iSoron** [comment]
Thanks everyone for voicing your opinion on this feature. In my understanding, this feature request is actually proposing a concrete solution (namely, display averages in measurable habits) to tackle the issue that Loop is not currently able to track weight, sleep or finances very well. As discussed in #2215, I'm not convinced that we should expand the scope of the app to cover these applications, as there are already very mature apps dedicated to them. Also note that properly tracking weight, sleep and finances requires way more than simply displaying averages; the solution proposed here, in my opinion, is very incomplete at best, and would provide a vastly inferior solution compared to these other apps.

**Reply by xdevs23** [reply]
I understand what you mean and where you're coming from, however, the PR, as it stands, is exactly what I need and it appears that I'm not the only one on this. While it's true that weight, sleep and finances don't fit into what the app was originally intended for, I believe that adding this feature, as it currently stands, would certainly provide value to at least a portion of users.

You're suggesting one should be using a different app for the purpose but I disagree with this stance for the following reasons:
* Convenience of doing all tracking in one app
* Easy to read and visually pleasing interface of this app
* Having already entered vast amounts of data into this app, making migration painful
* I simply do not want to use two different apps
* Merging the PR would suffice to add the only thing that is missing in my opinion

**Reply by Janssenhidal** [reply]
Hi @iSoron , thanks for sharing your perspective. I understand the concern about scope creep and that apps like Loop weren’t originally designed as full-fledged weight/sleep/finance trackers.

That said, I think this feature still fits naturally into Loop’s existing philosophy:

It doesn’t expand scope beyond habits, it just adds another way to view existing numerical data (totals vs averages). Users already enter numbers, and averages are just another aggregation of the same data.

**Keeps Loop flexible without being bloated.** Users can choose whether to use totals or averages, depending on the habit. No new complex workflows, just a small optional toggle.

**Helps with habits that Loop already supports.** Even if weight/sleep/finance aren’t the core use case, many “pure” habit use cases also benefit from averages (e.g. tracking average steps walked, average minutes meditated, average liters of water consumed).

**Low maintenance cost.** The code changes don’t require major redesigns. The feature is opt-in and doesn’t affect users who don’t need it.

I agree with you that full-on weight/sleep/finance tracking requires specialized apps, but in my view, having a simple average option doesn’t try to replace those apps, it just makes Loop more useful to those who are already using it for such habits.

**Reply by djvanderlaan** [reply]
I understand you want to prevent scope creep, but as the other also indicate, I believe this falls within the current scope of habits. 

> Loop is a mobile app that helps you create and maintain good habits, allowing you to achieve your long-term goals. Detailed graphs and statistics show you how your habits improved over time

Sleeping well, keeping weight within are also good habits, or at least indicators of good habits. And yes, there are probably other apps specifically for this, but the nice thing about habits is that it is simple and gives an overview of all habits in one place. And its simplicity. For running, I can use a Strava-like app that keeps track of all runs, the speed every minute, heartbeat (if I would have a device for that). But for me just knowing if I do the 2 runs a week minimum is enough. 

As @Janssenhidal also mentions, even for habits that are 'additive' like number of hours of sleep, I personally would find the average number of hours of sleep per day in the past month more useful than the total number of hours of sleep that month (Is 192 hours good/bad? How does that compare to the 51 hours this week? Am I improving?)

**Reply by Janssenhidal** [reply]
Since we haven't heard anything back from you @iSoron, shall we assume that you have closed off your mind to the possibility of having this feature added into the app?

**Reply by stigfaerch** [reply]
A couple of days ago I found this great app, after searching for an app where I can track habits and rate on statements about myself.
Statements can for example be something like:
"Yesterday was a great day!"
or
"This morning I feel fresh and well-rested"
And then I would rate 1-10.
But then I found that the values are just summed, which makes it less usable.

For me it is important that the app is open source and free. And as others have said, they really like to have these similar features in the same app and to not have to use closed-source proprietary apps.

@iSoron 
I'm thinking if there could be some way to find a good solution, which doesn't increase the scope of the core app?
For example if uhabits could be used as a base for another app which extends on uhabits, but an app that you don't have to maintain. Or maybe if it was possible to extend the functionality by adding another app?

**Reply by herrbrixter** [reply]
For habits like running, an average is far easier to understand than a sum. For example,

"I ran 4.5 km per day on average this month."

is a lot easier to understand and visualize than

"I ran 135 km total this month."

Runners don't talk about how much they run per month. They talk about their daily run.

**Reply by Glint-Eye** [reply]
I agree that an average Option in the history still has a lot of reasonable use cases that make sense for habit tracking.
For example: a typical Daily "Yes or No" Habit could be "Did I eat some Fruit Today". If i set a goal of for example 5 times per week, if i want to evaluate the development of this habit through bigger time frames, like monthly, quarterly, yearly, it just makes sense to have have an weekly average readout instead of a sum. Like "How close did I get to my set goal last december?" This quick question i have for myself when i look back can only really be quickly answered by an average for that timeframe, like "How many days in a week on avarage did i actually eat fruit in December". The Score is too cryptic for me for a clear idea of development.
Also a big part of motivation is as well the social aspect of personal development. If I want to tell someone about any changes, what would i tell them? "I'm trying to eat fruit at least 5 times per week. Last december i had a score of 85% in reaching that goal". That neither tells me a lot, nor anybody else. The score is helpful in getting a very quick vibe for "how am i doing?" but it doesn't give me information i can communicate in any reasonable way.
